### PR TITLE
Refactor and optimize multi-hop proof generation

### PR DESCRIPTION
**Pull request by dshiell**
Refactors and optimizes the multi-hop proof querying logic:
- removes inline verification
- fixes some subtle bugs
- updates/improves naming and comments

### CODE REVIEW

File: modules/core/multihop/proof_gen.go [hunk 1]
```diff
@@ -30,7 +28,7 @@ type Endpoint interface {
 
 	// QueryMinimumConsensusHeight returns the minimum height within the provided range at which the consensusState exists (processedHeight)
 	// and the height of the corresponding consensus state (consensusHeight).
```
Comment (7AC) [inline modules/core/multihop/proof_gen.go:30]: corresponding consensus state height?

File: modules/core/multihop/proof_gen.go [hunk 2]
```diff
@@ -64,53 +62,34 @@ func (p ChanPath) GetConnectionHops() []string {
...
-	defer func() {
-		if r := recover(); r != nil {
-			result = nil
-			err = sdkerrors.Wrapf(channeltypes.ErrMultihopProofGeneration, "%v", r)
-		}
-	}()
-
-	result = &channeltypes.MsgMultihopProofs{}
+// GenerateProof generates a proof for the given key on the the source chain, which is to be verified on the dest chain.
+func (p ChanPath) GenerateProof(key []byte, proofHeight exported.Height) (multihopProof channeltypes.MsgMultihopProofs, err error) {
```
Comment (7AC) [inline modules/core/multihop/proof_gen.go:67]: Should we take care of the linter errors?

File: modules/core/multihop/proof_gen.go [hunk 3]
```diff
@@ -121,96 +100,90 @@ func (p ChanPath) source() Endpoint {
...
+		chain,
+		host.FullConsensusStateKey(chain.ClientID(), consensusHeight),
 		bzConsStateAB,
-		processedHeight,
-		consStateBCRoot,
-		true,
+		proofHeight.Increment(), // need to match height used in QueryStateAtHeight
 	)
+
+	// debug code
```
Comment (7AC) [inline modules/core/multihop/proof_gen.go:162]: remove?

File: modules/core/multihop/proof_gen.go [hunk 4]
```diff
@@ -64,53 +62,34 @@ func (p ChanPath) GetConnectionHops() []string {
...
-
-	result = &channeltypes.MsgMultihopProofs{}
+// GenerateProof generates a proof for the given key on the the source chain, which is to be verified on the dest chain.
+func (p ChanPath) GenerateProof(key []byte, proofHeight exported.Height) (multihopProof channeltypes.MsgMultihopProofs, err error) {
 
 	// generate proof for key on source chain at the minimum consensus height known on the counterparty chain
-	maxProofHeight := p.source().Counterparty().QueryMaximumProofHeight(key, proofHeight, nil)
-	_, consensusHeightAB, err := p.source().Counterparty().QueryMinimumConsensusHeight(proofHeight, maxProofHeight)
+	maxProofHeight := p.source().QueryMaximumProofHeight(key, proofHeight, nil)
+	processedHeight, consensusHeight, err := p.source().Counterparty().QueryMinimumConsensusHeight(proofHeight, maxProofHeight)
```
Comment (7AC) [inline modules/core/multihop/proof_gen.go:70]: is processedHeight a chain height? If so maybe call it that to make it clearer (unless "processed" is standard)

**Review by 7AC — APPROVED** [review]
Looks good, just minor comments

**Comment by dshiell** [comment]
Noticed a few more things I should clean up.

**Pull request merged**